src/Nonlinear/ReverseAD/Coloring/Coloring.jl

```diff
@@ -0,0 +1,554 @@
+            end
+            if revmap[v] == 0
+                push!(vlist, v)
+                revmap[v] = length(vlist)
+                adjcount[v] = 0
+            end
+            adjcount[u] += 1
+            adjcount[v] += 1
+        end
+        # set up offsets for adjlist
+        offset = Vector{Int}(undef, length(vlist) + 1)
+        offset[1] = 1
+        for k in 1:length(vlist)
+            offset[k+1] = offset[k] + adjcount[vlist[k]]
+            adjcount[vlist[k]] = 0
+        end
+        # adjlist for node u in twocolor idx starts at
+        # vec[offset[u]]
+        # u has global index vlist[u]
+        vec = Vector{Int}(undef, offset[length(vlist)+1] - 1)
+        # now fill in
+        for k in 1:length(my_edges)
+            u, v = my_edges[k]
+            u_rev = revmap[u] # indices in the subgraph
+            v_rev = revmap[v]
+            vec[offset[u_rev]+adjcount[u]] = v_rev
+            vec[offset[v_rev]+adjcount[v]] = u_rev
+            adjcount[u] += 1
+            adjcount[v] += 1
+        end
+        resize!(cmap, length(vlist))
+        order, parent =
+            reverse_topological_sort_by_dfs(vec, offset, length(vlist), cmap)
+        for k in 1:length(vlist)
+            revmap[vlist[k]] = 0  # clear for reuse
+        end
+        postorder[idx] = order
+        parents[idx] = parent
+        vertexmap[idx] = vlist
+    end
+    return RecoveryInfo(
+        vertexmap,
+        postorder,
+        parents,
+        color,
+        num_colors,
+        _num_edges(g),
+        local_indices,
+    )
+end
+
+_normalize(i, j) = (j > i) ? (j, i) : (i, j)
+
+function _indirect_recover_structure(rinfo::RecoveryInfo)
+    N = length(rinfo.color)
+    I = zeros(Int, rinfo.nnz + N)
+    J = zeros(Int, rinfo.nnz + N)
+    k = 0
+    for i in 1:N
+        k += 1
+        I[k] = J[k] = i
+    end
+    for t in 1:length(rinfo.postorder)
+        vmap = rinfo.vertexmap[t]
+        order = rinfo.postorder[t]
+        parent = rinfo.parents[t]
+        for z in 1:length(order)
+            v = order[z]
+            p = parent[v]
+            if p == 0
+                continue
+            end
+            k += 1
+            I[k], J[k] = _normalize(vmap[v], vmap[p])
+        end
+    end
+    @assert k == rinfo.nnz + N
+    return I, J
+end
+
+"""
+    hessian_color_preprocess(
+        edgelist,
+        num_total_var,
+        seen_idx = IndexedSet(0),
+    )
+
+`edgelist` contains the nonzeros in the Hessian, *including* nonzeros on the
+diagonal.
+"""
+function hessian_color_preprocess(
+    edgelist,
+    num_total_var,
+    seen_idx = IndexedSet(0),
+)
+    resize!(seen_idx, num_total_var)
+    I, J = Int[], Int[]
+    for (i, j) in edgelist
+        push!(seen_idx, i)
+        push!(seen_idx, j)
+        push!(I, i)
+        push!(J, j)
+    end
+    local_indices = sort!(seen_idx.nzidx[1:seen_idx.nnz])
+    empty!(seen_idx)
+    global_to_local_idx = seen_idx.nzidx # steal for storage
```

Comment:
"steal for storage" is not clear?

Reply:
We can refactor Coloring in a different PR and document exactly what it does in more detail.